nibabel/streamlines/tck.py

```diff
@@ -0,0 +1,440 @@
+            raise HeaderError("Missing 'datatype' attribute in TCK header.")
+
+        if not hdr['datatype'].startswith('Float32'):
+            msg = ("TCK only supports float32 dtype but 'dataype: {}' was"
+                   " specified in the header.").format(hdr['datatype'])
+            raise HeaderError(msg)
+
+        if 'file' not in hdr:
+            raise HeaderError("Missing 'file' attribute in TCK header.")
+
+        if hdr['file'].split()[0] != '.':
+            msg = ("TCK only supports single-file - in other words the"
+                   " filename part must be specified as '.' but '{}' was"
+                   " specified.").format(hdr['file'].split()[0])
+            raise HeaderError("Missing 'file' attribute in TCK header.")
+
+        # Set endianness and _dtype attributes in the header.
+        hdr[Field.ENDIANNESS] = '<'
+        if hdr['datatype'].endswith('BE'):
+            hdr[Field.ENDIANNESS] = '>'
+
+        hdr['_dtype'] = np.dtype(hdr[Field.ENDIANNESS] + 'f4')
+
+        # Keep the file position where the data begin.
+        hdr['_offset_data'] = int(hdr['file'].split()[1])
+
+        # Set the file position where it was, if it was previously open.
+        if start_position is not None:
+            fileobj.seek(start_position, os.SEEK_SET)
+
+        return hdr
+
+    @staticmethod
+    def _read(fileobj, header, buffer_size=4):
+        """ Return generator that reads TCK data from `fileobj` given `header`
+
+        Parameters
+        ----------
+        fileobj : string or file-like object
+            If string, a filename; otherwise an open file-like object
+            pointing to TCK file (and ready to read from the beginning
+            of the TCK header). Note that calling this function
+            does not change the file position.
+        header : dict
+            Metadata associated with this tractogram file.
+        buffer_size : float, optional
+            Size (in Mb) for buffering.
+
+        Yields
+        ------
+        points : ndarray of shape (n_pts, 3)
+            Streamline points
+        """
+        dtype = header["_dtype"]
+        coordinate_size = 3 * dtype.itemsize
+        # Make buffer_size an integer and a multiple of coordinate_size.
+        buffer_size = int(buffer_size * MEGABYTE)
+        buffer_size += coordinate_size - (buffer_size % coordinate_size)
+
+        with Opener(fileobj) as f:
+            start_position = f.tell()
+
+            # Set the file position at the beginning of the data.
+            f.seek(header["_offset_data"], os.SEEK_SET)
+
+            eof = False
+            buff = b""
+            pts = []
+
+            i = 0
+
+            while not eof or not np.all(np.isinf(pts)):
+
+                if not eof:
+                    bytes_read = f.read(buffer_size)
+                    buff += bytes_read
+                    eof = len(bytes_read) == 0
+
+                # Read floats.
+                pts = np.frombuffer(buff, dtype=dtype)
```

Comment:
Only a small portion will get re-casted. It would correspond to the last sequence of points that is partially contained in `buff`.
In short, the code is doing the following.
1. Read a chunk of data from the file (a buffer).
2. Determine how many points we've read (i.e. nb. of triples [NaN, NaN, NaN])
3.Yield points, one streamline at the time.
4. Flush the beginning of the buff (that is up to the last NaN triple we've detected).
5. Goto step 1 (making sure we concatenate the new buffer to what is remaining of the current buffer) until we detect a Inf triple (or there is nothing to read).